### PR TITLE
#25 - Assign user to organizations on login

### LINKED ISSUE #25: Assign user to organizations on login
Check what organizations user is in, and assign him accordingly when he is logging

### PR DESCRIPTION
**Pull request by orlowski11**
This should close #25

Created service and tests for assigning users to organizations

### CODE REVIEW

File: tests/Unit/ExampleTest.php
Comment (EwelinaSkrzypacz) [inline tests/Unit/ExampleTest.php:1]: Do we really need this empty file?

File: tests/Feature/AssignUserToOrganizationsTest.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+        Http::fake([
+            "https://api.github.com/user/orgs" => Http::response([
+                ["id" => 123, "login" => "org1"],
+            ], 200),
+        ]);
+
+        $organization = Organization::factory()->create(["github_id" => 123]);
+
+        (new AssignUserToOrganizationsService())->assign();
```
Comment (EwelinaSkrzypacz) [inline tests/Feature/AssignUserToOrganizationsTest.php]:
```suggestion
        new AssignUserToOrganizationsService()->assign();
```

Comment (EwelinaSkrzypacz) [reply tests/Feature/AssignUserToOrganizationsTest.php]: I think we can delete these parenthesis

File: tests/Feature/AssignUserToOrganizationsTest.php
Comment (EwelinaSkrzypacz) [inline tests/Feature/AssignUserToOrganizationsTest.php:1]: And how about not happy path scenarios? Like what happens when GitHub API returns 500? Or when user is not login?

File: app/Services/AssignUserToOrganizationsService.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace App\Services;
+
+use App\Models\Organization;
+use App\Models\User;
+use Illuminate\Support\Facades\Http;
+
+class AssignUserToOrganizationsService
+{
+    public function assign(): void
```
Comment (kamilpiech97) [inline app/Services/AssignUserToOrganizationsService.php]:
```suggestion
    public function assign(User $user): void
```

Comment (kamilpiech97) [reply app/Services/AssignUserToOrganizationsService.php]: Lets use `user` from `callback` method from `GithubController`

File: app/Services/AssignUserToOrganizationsService.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        $user = User::query()->where("id", auth()->user()->id)->first();
+
+        $response = Http::withHeaders([
+            "Authorization" => "Bearer " . $user->github_token,
+        ])->get("https://api.github.com/user/orgs");
```
Comment (kamilpiech97) [inline app/Services/AssignUserToOrganizationsService.php]: What about handling error response or non-functional service errors?
Maybe move that action to new service - good package for that: https://docs.saloon.dev/

File: database/factories/OrganizationFactory.php [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+use Illuminate\Database\Eloquent\Factories\Factory;
+use Illuminate\Support\Str;
+
+class OrganizationFactory extends Factory
+{
+    public function definition(): array
+    {
+        return [
+            "name" => fake()->name(),
+            "github_id" => rand(),
```
Comment (kamilpiech97) [inline database/factories/OrganizationFactory.php]: Maybe use faker method?

File: database/factories/UserFactory.php [hunk 5]
```diff
@@ -18,16 +19,11 @@ public function definition(): array
         return [
             "name" => fake()->name(),
             "email" => fake()->unique()->safeEmail(),
-            "email_verified_at" => now(),
-            "password" => "$2y$10$92IXUNpkjO0rOQ5byMi.Ye4oKoEa3Ro9llC/.og/at2.uheWG/igi",
+            "github_id" => rand(),
```
Comment (kamilpiech97) [inline database/factories/UserFactory.php]: Maybe use faker method?

File: tests/Feature/AssignUserToOrganizationsTest.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+use Illuminate\Support\Facades\Http;
+use Tests\TestCase;
+
+class AssignUserToOrganizationsTest extends TestCase
+{
+    use RefreshDatabase;
+
+    public function testAssignFunctionWithValidResponse(): void
+    {
+        $user = User::factory()->create();
```
Comment (kamilpiech97) [inline tests/Feature/AssignUserToOrganizationsTest.php]: Maybe move creatig user to `setUp` method, u are doing same thing many times.

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

File: tests/Feature/AssignUserToOrganizationsTest.php (hunk 6 above)
Comment (orlowski11) [reply tests/Feature/AssignUserToOrganizationsTest.php]: moved creating user and service to setup

File: database/factories/UserFactory.php (hunk 5 above)
Comment (orlowski11) [reply database/factories/UserFactory.php]: changed to faker methods

File: database/factories/OrganizationFactory.php (hunk 4 above)
Comment (orlowski11) [reply database/factories/OrganizationFactory.php]: changed to faker methods

File: app/Services/AssignUserToOrganizationsService.php (hunk 3 above)
Comment (orlowski11) [reply app/Services/AssignUserToOrganizationsService.php]: used saloon for sending request and error response handling

File: app/Services/AssignUserToOrganizationsService.php (hunk 2 above)
Comment (orlowski11) [reply app/Services/AssignUserToOrganizationsService.php]: changed assign to use user from callback method

File: tests/Feature/AssignUserToOrganizationsTest.php
Comment (orlowski11) [reply tests/Feature/AssignUserToOrganizationsTest.php:1]: created tests for error responses. assign service now uses user from callback method, so i dont know if testing if user is logged in is needed?

File: tests/Feature/AssignUserToOrganizationsTest.php (hunk 1 above)
Comment (orlowski11) [reply tests/Feature/AssignUserToOrganizationsTest.php]: moved service creation to setup

File: tests/Unit/ExampleTest.php
Comment (orlowski11) [reply tests/Unit/ExampleTest.php:1]: tests in workflow failed with empty unit folder, thats why i put the example test back

File: app/Services/AssignUserToOrganizationsService.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public function __construct(
+        protected GithubConnector $githubConnector,
+    ) {}
+
+    public function assign(User $user): void
+    {
+        $request = new GetUsersOrganizationsRequest($user);
+
+        $response = $this->githubConnector->send($request);
```
Comment (kamilpiech97) [inline app/Services/AssignUserToOrganizationsService.php]: U can use https://docs.saloon.dev/digging-deeper/data-transfer-objects for handling data from response

File: app/Services/AssignUserToOrganizationsService.php (hunk 3 above)
Comment (kamilpiech97) [reply app/Services/AssignUserToOrganizationsService.php]: Where are u handling errors?

File: tests/Unit/ExampleTest.php
Comment (EwelinaSkrzypacz) [inline tests/Unit/ExampleTest.php:1]: I don't think we need this empty file

**Comment by EwelinaSkrzypacz** [comment]
Conflicts :(

**Review by kamilpiech97 — APPROVED** [review]

**Pull request merged**